### PR TITLE
Move runtime details into ContainerRuntime

### PR DESCRIPTION
**Pull request by bart0sh**
**What this PR does / why we need it**:

This PR moves all docker and CRI specifics into new ContainerRuntime
entity to isolate container runtime details from the rest of the code.

This should help to avoid having CRI and docker specific checks and code all over the place.

**NOTE: this is a not a finished work, but rather an RFC.**

Things to do:
- test manually in docker and cri-o environments

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/cmd/config.go
Comment (kad) [inline cmd/kubeadm/app/cmd/config.go]: I'd rather call this kind of variables `runtime` or `containerRuntime`.

File: cmd/kubeadm/app/cmd/reset.go
Comment (kad) [inline cmd/kubeadm/app/cmd/reset.go]: same here, for readability of code, better `containerRuntime`

File: cmd/kubeadm/app/util/BUILD
Comment (kad) [inline cmd/kubeadm/app/util/BUILD]: I think better to call it just `runtime.go`

File: cmd/kubeadm/app/util/cruntime.go
Comment (kad) [inline cmd/kubeadm/app/util/cruntime.go]: `runtimeType` ?

File: cmd/kubeadm/app/util/cruntime.go
Comment (kad) [inline cmd/kubeadm/app/util/cruntime.go]: better to do here `switch` with `cruntime.rtype == CRI', `cruntime.rtype == Docker` and `defaults:` with error.

File: cmd/kubeadm/app/util/cruntime.go
Comment (kad) [inline cmd/kubeadm/app/util/cruntime.go]: same here. use `switch` with known runtimes and `defaults:` if runtime is unknown.

**Review by kad — COMMENTED** [review]
@bart0sh please add tests

/ok-to-test
/cc @luxas @timothysc @fabriziopandini

**Comment by bart0sh** [comment]
@kad thank you for the review. Updated the code as you've suggested.

**Review by luxas — COMMENTED** [review]
Let me get back to this next week or so, sorry I don't have to review at the moment. Maybe @timothysc, @fabriziopandini or @chuckha have time?

That said, I don't want us to target v1.11 with this as we have other major refactors ongoing which change behavior and are more significant that we need to get into v1.11 to meet our goals for this year. This seems like a good direction, but isn't directly user visible, so it can wait until early next cycle to get merged. Makes sense to others?

**Comment by bart0sh** [comment]
@luxas It makes sense to me. This is not for 1.11 for sure.

**Comment by timothysc** [comment]
punting to 1.12 on this one, we're one day out from code-freeze atm.

**Comment by bart0sh** [comment]
@kad, @luxas 
status report:
- write test cases for the ContainerRuntime - done
- fix failing test cases for reset, preflight, etc - done

all: please, review

**Comment by bart0sh** [comment]
@luxas > This seems like a good direction, but isn't directly user visible

Thanks. Although I agree with you about not targeting 1.11 I'd argue that this **is** user visible change. Here is a list of bugs caused by the current state of CRI "support" in kubeadm:
- https://github.com/kubernetes/kubeadm/issues/657
- https://github.com/kubernetes/kubeadm/issues/749
- https://github.com/kubernetes/kubeadm/issues/814
- https://github.com/kubernetes/kubeadm/issues/811
- https://github.com/kubernetes/kubeadm/issues/285

And a fraction of countless attempts to fix it:
- https://github.com/kubernetes/kubernetes/pull/63297
- https://github.com/kubernetes/kubernetes/pull/58802
- https://github.com/kubernetes/kubernetes/pull/55055
- https://github.com/kubernetes/kubernetes/pull/62481
...

This PR is aiming to improve the situation by isolating runtime specifics in ContainerRuntime API. This should improve current situation or at least start this process.

Thoughts?

File: cmd/kubeadm/app/preflight/checks.go
Comment (neolit123) [inline cmd/kubeadm/app/preflight/checks.go]: please remove this extra whitespace.

File: cmd/kubeadm/app/util/runtime_test.go
Comment (neolit123) [inline cmd/kubeadm/app/util/runtime_test.go]: i think you can refactor this function to not use `i` and `num` and instead:
```
for _, action := range fcmd.RunScript {
    ...
}
```

**Review by neolit123 — COMMENTED** [review]
added some comments. i think i'm OK with the wrapping of the CRI here. some of the recent additions in the file `utils/cgroupdriver.go` can be unified with this eventually.

File: cmd/kubeadm/app/util/runtime_test.go
Comment (bart0sh) [reply cmd/kubeadm/app/util/runtime_test.go]: Thank you for pointing it out. I'll fix that.

File: cmd/kubeadm/app/util/runtime_test.go
Comment (bart0sh) [reply cmd/kubeadm/app/util/runtime_test.go]: I don't think it's possible as there is at least one case where fcmd.CombinedOutputScript is passed to this function.

File: cmd/kubeadm/app/util/runtime.go
Comment (neolit123) [reply cmd/kubeadm/app/util/runtime.go]: yes, but until then it's always `nil`. my point is to remove the `rcode == nil` check.

File: cmd/kubeadm/app/util/runtime_test.go
Comment (neolit123) [reply cmd/kubeadm/app/util/runtime_test.go]: can you point out the case where `len()` is needed, please?

Comment (bart0sh) [reply cmd/kubeadm/app/util/runtime_test.go]: sure:
```
$ grep genFakeActions runtime_test.go 
func genFakeActions(fcmd *fakeexec.FakeCmd, num int) []fakeexec.FakeCommandAction {
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.CombinedOutputScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
		CommandScript: genFakeActions(&fcmd, len(fcmd.RunScript)),
```
In most cases your suggestion to use fcmd.RunScript would work, but it wouldn't work when generating actions based on fcmd.CombinedOutputScript. Am I missing something?

File: cmd/kubeadm/app/util/runtime.go
Comment (bart0sh) [reply cmd/kubeadm/app/util/runtime.go]: if I remove it it will be reassigned to the next error if it occurs. I want it to be assigned to the first one.

Comment (neolit123) [reply cmd/kubeadm/app/util/runtime.go]: aaah, it preserves a state across loop iterations, i see.
it's OK in that case.

File: cmd/kubeadm/app/util/runtime_test.go
Comment (neolit123) [reply cmd/kubeadm/app/util/runtime_test.go]: given other fields are used like `fcmd.CombinedOutputScript` to fetch a length, then `genFakeActions()` is better, as is.
agreed.

File: cmd/kubeadm/app/util/runtime.go
Comment (chuckha) [inline cmd/kubeadm/app/util/runtime.go]: What do you think about having one case statement in the constructor that creates some new struct (DockerInterface or CrictlInterface) that implements an interface which contains the CRI functions needed for kubeadm?

This would reduce the case statements in each function and call the Interface instead. This makes it easier to mock in tests and pushes the implementation out of the ContainerRuntime to be isolated in some struct. Adding new runtimes would also be easier but I admit that will likely not happen any time soon.

These functions could shrink down to `return runtime.CRI.List()`, `return runtime.CRI.Rm(containers...)` or something similar.

Comment (bartosh) [reply cmd/kubeadm/app/util/runtime.go]: @chuckha Will do. Thank you for the review!

File: cmd/kubeadm/app/cmd/reset.go
Comment (chuckha) [reply cmd/kubeadm/app/cmd/reset.go]: +1, glog.V(x) for anything that is not required output

File: cmd/kubeadm/app/util/runtime.go
Comment (luxas) [inline cmd/kubeadm/app/util/runtime.go]: This automatically is restricted to the pods run by kubernetes? Just so we don't list anything we don't own here, right?

File: cmd/kubeadm/app/util/runtime.go
Comment (luxas) [inline cmd/kubeadm/app/util/runtime.go]: I'd rename this `ListKubeContainers()` so it's clear that this is specifically about the Kubernetes-managed containers

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: Above we hard-code `"/var/run/dockershim.sock"`, can you change so it's using the constant from the API?

File: cmd/kubeadm/app/cmd/reset_test.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset_test.go]: improve this unit test with more test cases?

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: can inline this err in the if clause

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: make this return an error

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: return the error

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: return the error

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: return the error

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [reply cmd/kubeadm/app/cmd/reset.go]: add a comment please

File: cmd/kubeadm/app/cmd/reset.go
Comment (luxas) [inline cmd/kubeadm/app/cmd/reset.go]: if I'm reading this right, we used to try to use crictl for removing docker containers earlier (given crictl existed in PATH)
What's changing is that we now don't try to do that, but always use docker for dockershim.sock
The other change is that in case of an other CRI runtime, don't fallback on docker.
Is that right? Both of these changes SGTM

**Review by luxas — COMMENTED** [review]
Thanks very much @bart0sh for the change, this overall
looks like a great refactor to me :+1:! I'm okay with the direction of this PR.

I didn't look much on the unit testing yet, deferring to other reviewers e.g. @chuckha @neolit123 and @timothysc for that. However, reviewed all of the code and provided the initial feedback.

File: cmd/kubeadm/app/cmd/reset.go
Comment (bart0sh) [reply cmd/kubeadm/app/cmd/reset.go]: Yes, that's correct. I tried to make it more explicit and only use crictl when user specifies CRI socket, otherwise use docker. Theoretically it would probably be even better if we add new command line option for that (--use-cri), but I didn't want to do that in this PR as it's already quite big.

File: cmd/kubeadm/app/cmd/reset.go
Comment (bart0sh) [reply cmd/kubeadm/app/cmd/reset.go]: This is addressed in this PR I believe: https://github.com/kubernetes/kubernetes/pull/64911

File: cmd/kubeadm/app/cmd/config.go
Comment (bart0sh) [reply cmd/kubeadm/app/cmd/config.go]: criutil sounds like something only about CRI. Would it be better to call it utilruntime or something similar?

File: cmd/kubeadm/app/cmd/reset_test.go
Comment (bart0sh) [reply cmd/kubeadm/app/cmd/reset_test.go]: done. 100% coverage

**Comment by bart0sh** [comment]
@lucas thank you for the review! It was very useful and educational.

File: cmd/kubeadm/app/preflight/checks.go
Comment (chuckha) [reply cmd/kubeadm/app/preflight/checks.go]: Look at the SystemVerification check; it will add a docker check if we are using docker, but the way it determines that is `if sysver.CRISocket == kubeadmdefaults.DefaultCRISocket {` which should be cleaned up in this PR

**Review by chuckha — COMMENTED** [review]
This is looking great, thanks for being patient while we got through 1.11!

Just one cleanup and then it looks good to me

File: cmd/kubeadm/app/preflight/checks.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: @chuckha hmm, still don't get it. Are you talking about this change?
```
@@ -532,8 +525,11 @@ func (sysver SystemVerificationCheck) Check() (warnings, errors []error) {
 	var validators = []system.Validator{
 		&system.KernelValidator{Reporter: reporter}}
 
-	// run the docker validator only with dockershim
-	if sysver.CRISocket == kubeadmapiv1alpha3.DefaultCRISocket {
+	// run the docker validator only with docker runtime
+	containerRuntime, err := utilruntime.NewContainerRuntime(utilsexec.New(), sysver.CRISocket)
+	if err != nil {
+		errs = append(errs, err)
+	} else if containerRuntime.IsDocker() {
 		// https://github.com/kubernetes/kubeadm/issues/533
 		validators = append(validators, &system.DockerValidator{Reporter: reporter})
 	}
```

What's wrong with it?

**Comment by bart0sh** [comment]
@chuckha thank you for the review and encouragement!

File: cmd/kubeadm/app/preflight/checks.go
Comment (bart0sh) [reply cmd/kubeadm/app/preflight/checks.go]: @luxas ah, got it at last! will do. thank you for the suggestion!

**Comment by bart0sh** [comment]
Worked around this issue: https://github.com/kubernetes/kubeadm/issues/928 
```
+		// !!! temporary work around crictl warning:
+		// Using "/var/run/crio/crio.sock" as endpoint is deprecated,
+		// please consider using full url format "unix:///var/run/crio/crio.sock"
+		if filepath.IsAbs(criSocket) {
+			if runtime.GOOS != "windows" {
+				criSocket = "unix://" + criSocket
+			}
+		}
```

Without this work around crictl produces warning and that causes issues like this one: https://github.com/kubernetes/kubeadm/issues/926

File: cmd/kubeadm/app/preflight/checks_test.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks_test.go]: should be v1alpha3

File: cmd/kubeadm/app/util/runtime/runtime.go
Comment (luxas) [inline cmd/kubeadm/app/util/runtime/runtime.go]: should be v1alpha3

File: cmd/kubeadm/app/preflight/checks.go
Comment (luxas) [inline cmd/kubeadm/app/preflight/checks.go]: :+1:

File: cmd/kubeadm/app/util/runtime/runtime.go
Comment (luxas) [inline cmd/kubeadm/app/util/runtime/runtime.go]: Please hold this PR until the other PR you submitted to fix this is merged

File: cmd/kubeadm/app/util/runtime/runtime.go
Comment (luxas) [inline cmd/kubeadm/app/util/runtime/runtime.go]: I'd name this runtime, and the `"runtime"` import e.g. `"goruntime"`

File: cmd/kubeadm/app/util/runtime/runtime.go
Comment (luxas) [inline cmd/kubeadm/app/util/runtime/runtime.go]: `return errors.NewAggregate(errs)` where errors is `"k8s.io/apimachinery/pkg/util/errors"`

**Review by luxas — APPROVED** [review]
This looks really great!
I left a couple of smaller review comments, please wait for the other PR, rebase to use v1alpha3 that recently got merged, and then this LGTM :)

**Review by timothysc — APPROVED** [review]
/approve 

will let others who have been on this thread lgtm.

**Comment by neolit123** [comment]
@bart0sh 

hi, please have a look at this failing CI test:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/64611/pull-kubernetes-verify/98814/

it just needs a small formatting fix.

**Comment by bart0sh** [comment]
hm, it's approved, but still not merged. i guess it still requires lgtm. Can anybody help?

**Comment by neolit123** [comment]
thank you @bart0sh , seems good to me.
waiting on @luxas for the final LGTM.

File: cmd/kubeadm/app/util/runtime/runtime_test.go
Comment (chuckha) [inline cmd/kubeadm/app/util/runtime/runtime_test.go]: Best practice here is to run each test case in a sub test. https://golang.org/pkg/testing/#T.Run

The output is much nicer and we know exactly which test case is failing. Right now the output would be confusing and we wouldn't know which of the test cases were failing.

**Review by chuckha — CHANGES_REQUESTED** [review]
If you wrap all the tests in a t.Run() this lgtm

**Review by luxas — APPROVED** [review]
/lgtm
Based on both me not spotting anything strange and @chuckha's review. If you can fix the unit tests to use subtests before the merge, fine, if not just follow-up. But I want to get this in now. We don't use subtests in ~90% of the other unit tests yet either so...

**Pull request merged**
